### PR TITLE
mrc-2183 pass through error details even when using throwOnError

### PR DESCRIPTION
**Pull request by hillalex**
At the moment, if a `PorcelainAPI` instance has the `throwOnError` flag, a generic exception is raised. This PR changes this so that the error detail from the porcelain API is passed through as the exception message.

### CODE REVIEW

File: src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/OrderlyServerClient.kt [hunk 1]
```diff
@@ -18,7 +19,7 @@ interface OrderlyServerAPI : PorcelainAPI
 
 class OrderlyServerClient(
         config: Config,
-        client: OkHttpClient = OkHttpClient()
+        client: OkHttpClient = httpClient
```
Comment (hillalex) [inline src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/OrderlyServerClient.kt:22]: this is unrelated, but the `OkHttpClient` docs say it's best to use a singleton instance of this class throughout the app

File: src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/IntegrationTest.kt [hunk 2]
```diff
@@ -87,7 +87,7 @@ abstract class IntegrationTest
 
     protected fun assertJsonContentType(response: Response)
     {
-        Assertions.assertThat(response.headers["content-type"]).isEqualTo("application/json;charset=utf-8")
+        Assertions.assertThat(response.headers["content-type"]).contains("application/json")
```
Comment (hillalex) [inline src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/IntegrationTest.kt:90]: porcelain might return this type without charset

File: src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/unit_tests/controllers/api/OutpackControllerTests.kt [hunk 3]
```diff
@@ -8,37 +11,52 @@ import org.vaccineimpact.orderlyweb.PorcelainAPI
 import org.vaccineimpact.orderlyweb.PorcelainResponse
 import org.vaccineimpact.orderlyweb.controllers.api.OutpackController
 import org.vaccineimpact.orderlyweb.tests.unit_tests.controllers.api.ControllerTest
+import spark.RequestResponseFactory
+import spark.Response
+import java.io.OutputStream.nullOutputStream
+import javax.servlet.ServletOutputStream
+import javax.servlet.http.HttpServletResponse
 
-class ArtefactControllerTests : ControllerTest()
```
Comment (hillalex) [inline src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/unit_tests/controllers/api/OutpackControllerTests.kt:12]: old copy and paste error in the name here!

File: src/customConfigTests/src/test/kotlin/org/vaccineimpact/orderlyweb/customConfigTests/GithubAuthenticationTests.kt [hunk 4]
```diff
@@ -125,26 +126,22 @@ class GithubAuthenticationTests : CustomConfigTests()
 
     private fun assertAuthSuccess(result: Response)
     {
-        result.use {
```
Comment (hillalex) [inline src/customConfigTests/src/test/kotlin/org/vaccineimpact/orderlyweb/customConfigTests/GithubAuthenticationTests.kt:128]: this change is because I've changed this `Response` type so it's no longer closeable (instead the response from OkHttp gets closed when we create this: https://github.com/vimc/orderly-web/pull/508/files#diff-cd7bfd9025f911f7b6bc62e1104e07798a28f5ac57a55618b844b38da4fbce13L106)

File: src/testHelpers/src/main/kotlin/org/vaccineimpact/orderlyweb/test_helpers/http/HttpClient.kt [hunk 5]
```diff
@@ -29,7 +25,9 @@ object HttpClient
             .url(url)
             .headers(headers.toHeaders())
             .build()
-        return Response(client.newCall(request).execute())
+        client.newCall(request).execute().use {
```
Comment (hillalex) [inline src/testHelpers/src/main/kotlin/org/vaccineimpact/orderlyweb/test_helpers/http/HttpClient.kt:28]: The changes here are not related to the main code change, but we were still getting loads of "leaked connection" warnngs in the tests. this was because the `Response` class here needed to be closed in order to close the underlying `OkHttpResponse`, but this was basically just not happening almost anywhere. I refactored this so that all the response closing logic is contained hgere and individual tests don't have to worry about it (since they weren't anyway).

File: docs/spec/Error.schema.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
     "id": "Error",
     "properties": {
-        "error": { "$ref": "ErrorCode.schema.json" },
+        "error": { "type": "string" },
```
Comment (hillalex) [inline docs/spec/Error.schema.json:4]: porcelain doesn't restrict this

Comment (EmmaLRussell) [reply docs/spec/Error.schema.json:4]: Does that mean we can lose `ErrorCode.schema.json`?

File: src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/BundleController.kt [hunk 7]
```diff
@@ -19,23 +19,20 @@ class BundleController(
             this(
                     context,
                     AppConfig(),
-                    OrderlyServerClient(AppConfig()).throwOnError()
+                    OrderlyServerClient(AppConfig())
```
Comment (EmmaLRussell) [inline src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/BundleController.kt:22]: Don't really understand the changes to this file. Do we not need to specify throwOnError here because it's doing a pass through anyway?

File: src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/Controller.kt [hunk 8]
```diff
@@ -13,6 +13,15 @@ abstract class Controller(val context: ActionContext, val appConfig: Config = Ap
         return response.text
     }
 
+    protected fun writeResponseToOutputStream(response: PorcelainResponse): Boolean
```
Comment (EmmaLRussell) [inline src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/Controller.kt:16]: This is just a refactor to avoid repeated code?

File: src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/QueueController.kt [hunk 9]
```diff
@@ -18,7 +18,7 @@ class QueueController(
             this(
                     context,
                     AppConfig(),
-                    OrderlyServerClient(AppConfig()).throwOnError()
+                    OrderlyServerClient(AppConfig())
```
Comment (EmmaLRussell) [inline src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/QueueController.kt:21]: ..same for removing throwOnError here?

File: src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt [hunk 10]
```diff
@@ -33,7 +33,7 @@ class BundlePackTests : IntegrationTest()
     {
         val response = apiRequestHelper.post("/bundle/pack/minimal/", emptyMap(), ContentTypes.zip, fakeGlobalReportReviewer())
         assertThat(response.statusCode).isEqualTo(200)
-        assertThat(response.headers["content-type"]).isEqualTo("application/zip")
+        assertThat(response.headers["content-type"]).isEqualTo("application/octet-stream")
```
Comment (EmmaLRussell) [inline src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt:36]: Not sure why this change has happened?

**Review by EmmaLRussell — COMMENTED** [review]
Looks good, I mostly have questions!

So requests which *don't* throw on error may handle the error response however the controller wants - which might include a pass through of the Porcelain response, so you get the orderly server error detail anyway.. or might involve constructing some other type of response in OW. Is that right?

File: src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/BundleController.kt (hunk 7 above)
Comment (hillalex) [reply src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/BundleController.kt:22]: Sorry, yeah, exactly. In both the below methods the response from `orderly.server` is now passed along (including errors).

File: src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt (hunk 10 above)
Comment (hillalex) [reply src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/integration_tests/tests/api/BundleTests.kt:36]: Now simply passing through the response from `orderly.server` (including all headers) and this is what `orderly.server` returns for this endpoint.

File: src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/QueueController.kt (hunk 9 above)
Comment (hillalex) [reply src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/controllers/api/QueueController.kt:21]: Yeah, unclear to me why we ever had this here.

**Review by EmmaLRussell — APPROVED** [review]
